tsdb/meta.go

```diff
@@ -515,6 +523,16 @@ func (d *DatabaseIndex) DropSeries(keys []string) {
 	atomic.AddInt64(&d.stats.NumSeries, -nDeleted)
 }
 
+// Dereference removes all references to data within b and moves them to the heap.
+func (d *DatabaseIndex) Dereference(b []byte) {
+	d.mu.Lock()
```

Comment:
Probably needs to be an `RLock` so that a slow delete doesn't lock up the whole database.